### PR TITLE
mercurial: use working patch links

### LINKED ISSUE #127452: mercurial: patches give a HTTP 404 response
**Describe the bug**

The current [derivation for mercurial](https://github.com/NixOS/nixpkgs/blob/353c081de2fc7193afb9cc176e3e36b66400e93a/pkgs/applications/version-management/mercurial/default.nix#L20-L31) downloads 2 patches, both of them 404.

I realized this when I tried to build my system from scratch without substitors.

**To Reproduce**

1. Have a system that somehow depends on mercurial (not exactly sure if this is reproducible with just `nix-build mercurial` or similar, this was also on a content addressed machine, and reproducing the issue in a more isolated way would require another toolchain bootstrap which takes several days)
2. rebuild the system from scratch using `--option substituters ''`
3. Wait 
4. Eventually see errors like this:
   ```
   sudo nixos-rebuild switch -L --flake . --fallback --max-jobs 1 --cores 1 --keep-going
   [sudo] password for nmelzer: 
   building the system configuration...
   D10638.diff> trying https://phab.mercurial-scm.org/file/data/oymk4awh2dd7q6cwjbzu/PHID-FILE-bfcr7qrp5spg42wspxpd/D10638.diff
   D10638.diff>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
   D10638.diff>                                  Dload  Upload   Total   Spent    Left  Speed
   D10638.diff>   0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0
   D10638.diff> curl: (22) The requested URL returned error: 404
   D10638.diff> error: cannot download D10638.diff from any mirror
   error: builder for '/nix/store/micicvl7b56bv8cmhjmblyii6i5vkk8x-D10638.diff.drv' failed with exit code 1
   D10639.diff> trying https://phab.mercurial-scm.org/file/data/re4uqdhtknjiacx2ogwu/PHID-FILE-4m26id65dno5gzix2ngh/D10639.diff
   D10639.diff>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
   D10639.diff>                                  Dload  Upload   Total   Spent    Left  Speed
   D10639.diff>   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
   D10639.diff> curl: (22) The requested URL returned error: 404
   D10639.diff> error: cannot download D10639.diff from any mirror
   error: builder for '/nix/store/y7xg9i2bwi8k00glwbrarin1kr1wwx1f-D10639.diff.drv' failed with exit code 1
   ```

**Expected behavior**

mercurial builds without errors

**Screenshots**

See log output from reproduction. Also `curl`ing the URLs as in the derivation linked in the description result in 404.

**Additional context**

The exact commit the system was build from is nixos/nixpkgs/fa0326ce5233f7d592271df52c9d0812bec47b84, though as the same path URLs are used in nixos/nixpkgs/master, this is probably only partially relevant.

**Notify maintainers**

\cc @edolstra @lukegb

**Metadata**

nix shell nixpkgs#nix-info --option trusted-substituters https://cache.nixos.org -c nix-info -m
 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.122, NixOS, 21.11.20210613.fa0326c (Porcupine)`
 - multi-user?: `no`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.4pre20210615_e6150de`
 - channels(nmelzer): `"home-manager, nixpkgs-20.09pre225970.726cbe0d405"`
 - nixpkgs: `/home/nmelzer/.nix-defexpr/channels/nixpkgs`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
- mercurial
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by Artturin**
###### Motivation for this change
Closes #127452

###### Things done
took the patches from the Download Raw Diff links 
https://phab.mercurial-scm.org/D10638
https://phab.mercurial-scm.org/D10639

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/version-management/mercurial/default.nix [hunk 1]
```diff
@@ -20,13 +20,13 @@ in python3Packages.buildPythonApplication rec {
...
-      sha256 = "0mfi324is02l7cnd3j0gbmg5rpyyqn3afg3f73flnfwmz5njqa5f";
+      url = "https://phab.mercurial-scm.org/file/data/l7p2r4zcctcr3pzlybv2/PHID-FILE-bwjzxlz6sbegk3s4irik/D10638.diff";
+      sha256 = "07bhprh5zwl37snn59jsvx0mnp6s0n3hi0zl8zwbb20nlbvfbvfz";
     })
     # https://phab.mercurial-scm.org/D10639, fixes https://bz.mercurial-scm.org/show_bug.cgi?id=6514
     (fetchpatch {
-      url = "https://phab.mercurial-scm.org/file/data/re4uqdhtknjiacx2ogwu/PHID-FILE-4m26id65dno5gzix2ngh/D10639.diff";
-      sha256 = "0h5ilrd2x1789fr6sf4k1mcvxdh0xdyr94yawdacw87v3x12c8cb";
+      url = "https://phab.mercurial-scm.org/file/data/v53nhburhtkhpccyecei/PHID-FILE-6v34oll6r2gkqo4ng5nt/D10639.diff";
+      sha256 = "1azj90r4xc4ygdli0bv1d4yi0nl5dx0vbiri3ihlfzaykhqgl3w9";
```
Comment (SuperSandro2000) [inline pkgs/applications/version-management/mercurial/default.nix:29]:
```suggestion
      sha256 = "0h5ilrd2x1789fr6sf4k1mcvxdh0xdyr94yawdacw87v3x12c8cb";
```

File: pkgs/applications/version-management/mercurial/default.nix [hunk 2]
```diff
@@ -20,13 +20,13 @@ in python3Packages.buildPythonApplication rec {
   patches = [
     # https://phab.mercurial-scm.org/D10638, needed for below patch to apply
     (fetchpatch {
-      url = "https://phab.mercurial-scm.org/file/data/oymk4awh2dd7q6cwjbzu/PHID-FILE-bfcr7qrp5spg42wspxpd/D10638.diff";
-      sha256 = "0mfi324is02l7cnd3j0gbmg5rpyyqn3afg3f73flnfwmz5njqa5f";
+      url = "https://phab.mercurial-scm.org/file/data/l7p2r4zcctcr3pzlybv2/PHID-FILE-bwjzxlz6sbegk3s4irik/D10638.diff";
+      sha256 = "07bhprh5zwl37snn59jsvx0mnp6s0n3hi0zl8zwbb20nlbvfbvfz";
```
Comment (SuperSandro2000) [inline pkgs/applications/version-management/mercurial/default.nix:24]: That can likely also be reverted.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 127620` at d8b597e9 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>43 packages failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/494701519f6d41e62ec98118113aa311">bumpver</a></li>
    <li><a href="https://gist.github.com/c31f6c3de58c95f5b14742a139fc0973">cabal2nix</a></li>
    <li><a href="https://gist.github.com/27475ff1dd85e60b79b0519377918ad3">dep2nix</a></li>
    <li><a href="https://gist.github.com/98c53e0be91cf2a119d96adaa8600690">devpi-client</a></li>
    <li><a href="https://gist.github.com/c9d6be9c3ff9eb538d56f36c8aa03675">dune-release</a></li>
    <li><a href="https://gist.github.com/1546d70534ee7458eb57e287d14ba831">git-cinnabar</a></li>
    <li><a href="https://gist.github.com/d0a0c0ab1567b52086f923fcfd9a3f17">git-remote-hg</a></li>
    <li><a href="https://gist.github.com/7b5bf7534cddc2ac76f21d3bcdc4b633">haskell-language-server</a></li>
    <li><a href="https://gist.github.com/a6e863a9ebb4bd6a6df4c3650868eb99">haskellPackages.cabal2nix-unstable</a></li>
    <li><a href="https://gist.github.com/3bee0162dcf68d9b403a3daaa3fc82e8">haskellPackages.ghcide</a></li>
    <li><a href="https://gist.github.com/7b4913674db92d1ba21204cf7aaf8383">haskellPackages.haskell-language-server</a></li>
    <li><a href="https://gist.github.com/7b9322bbf69e52bde40f10cc76052cc0">haskellPackages.hls-brittany-plugin</a></li>
    <li><a href="https://gist.github.com/57b8de0ccb84eceb8f7a4d68de9507fd">haskellPackages.hls-class-plugin</a></li>
    <li><a href="https://gist.github.com/469f985f9c6d713e2dd74af6ca712c1c">haskellPackages.hls-eval-plugin</a></li>
    <li><a href="https://gist.github.com/f3e7dab28d469710149bf7506f637670">haskellPackages.hls-explicit-imports-plugin</a></li>
    <li><a href="https://gist.github.com/f8236a6bc90194e3ff0154c033be79d7">haskellPackages.hls-haddock-comments-plugin</a></li>
    <li><a href="https://gist.github.com/2be3b1441a635d1c255da34cf8975db3">haskellPackages.hls-hlint-plugin</a></li>
    <li><a href="https://gist.github.com/e5c3253d848bce3e7a76900d7f8a7ccb">haskellPackages.hls-retrie-plugin</a></li>
    <li><a href="https://gist.github.com/069bbb45736900f6b71fe8b84ce68561">haskellPackages.hls-splice-plugin</a></li>
    <li><a href="https://gist.github.com/dc0f0e03382a005b504ce53c08351cdc">haskellPackages.hls-stylish-haskell-plugin</a></li>
    <li><a href="https://gist.github.com/591b3ebafe5be8993c6cf29eb9124809">haskellPackages.hls-tactics-plugin</a></li>
    <li><a href="https://gist.github.com/018cd6d8d4b5e8fa7eccc54298da7acb">haskellPackages.hls-test-utils</a></li>
    <li><a href="https://gist.github.com/4498de1f0d50e9f4c4c5f10cc50f7f5c">haskellPackages.retrie</a></li>
    <li><a href="https://gist.github.com/be0c6c5264bb9aa191907c8165f8b352">hound</a></li>
    <li><a href="https://gist.github.com/50a91cdbdd0c237ec0c2d356132e78eb">hydra-unstable</a></li>
    <li><a href="https://gist.github.com/6688ebe27ebbe79a9824f101b967eb26">mbed-cli</a></li>
    <li><a href="https://gist.github.com/08f1f918d24318ad9e00f84f1ebff572">mercurial</a> (python38Packages.mercurial)</li>
    <li><a href="https://gist.github.com/80cefc9bebca528301a0740d1a940e31">mercurialFull</a></li>
    <li><a href="https://gist.github.com/8ba9b57ab29fb533a0e5f0907b2a9777">nbstripout</a></li>
    <li><a href="https://gist.github.com/85162408be5d2c2c25f7a8eb3fe42658">nix-prefetch-hg</a></li>
    <li><a href="https://gist.github.com/bc78e93ffe5ad2e678122b2fdaafe0d8">nix-prefetch-scripts</a></li>
    <li><a href="https://gist.github.com/87ffb44ade5516e7f761c189d5993376">nix-update-source</a></li>
    <li><a href="https://gist.github.com/e6492fba76e829d367845fec7a4d7ddd">patroni</a></li>
    <li><a href="https://gist.github.com/fe61de1097197a4bd2baf9d0fd6cf5c6">powerline</a> (python38Packages.powerline)</li>
    <li><a href="https://gist.github.com/e144dc85382c00290b7f5136c14080aa">python38Packages.hg-evolve</a></li>
    <li><a href="https://gist.github.com/5b73b18b7308e7c7497dca3ebececb2f">python38Packages.hglib</a></li>
    <li><a href="https://gist.github.com/39f0af7e04fcde63d4b8b8e927b738bb">ydiff</a> (python38Packages.ydiff)</li>
    <li><a href="https://gist.github.com/3cb2a615c75122964bfd555be7028bae">python39Packages.hg-evolve</a></li>
    <li><a href="https://gist.github.com/a1f104af4f09e1d17d3cecbf61502a65">python39Packages.hglib</a></li>
    <li><a href="https://gist.github.com/69c69df25375d9b51db5f70ba9a7d43d">python39Packages.mercurial</a></li>
    <li><a href="https://gist.github.com/0504ed72e55e75e9ee4f897acfc0f17d">python39Packages.powerline</a></li>
    <li><a href="https://gist.github.com/9f5b39382a24f43921230f96829cde58">python39Packages.ydiff</a></li>
    <li><a href="https://gist.github.com/7d2781199c1d36dc6c9ca5e2813007ac">sourcehut.hgsrht</a></li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement. Does something look off? Please file an [issue](https://github.com/rmcgibbo/r-rmcgibbo/issues) or reach out on IRC.

File: pkgs/applications/version-management/mercurial/default.nix (hunk 1 above)
Comment (Artturin) [reply pkgs/applications/version-management/mercurial/default.nix:29]:
```
nix-prefetch-url "https://phab.mercurial-scm.org/file/data/v53nhburhtkhpccyecei/PHID-FILE-6v34oll6r2gkqo4ng5nt/D10639.diff"
[0.0 MiB DL]
path is '/nix/store/yka0g62n515qh2k6y8lli8cx3vpclfiw-D10639.diff'
1azj90r4xc4ygdli0bv1d4yi0nl5dx0vbiri3ihlfzaykhqgl3w9
``` weird

Comment (AndersonTorres) [reply pkgs/applications/version-management/mercurial/default.nix]: `fetchpatch` is different from `fetchurl`. `fetchpatch` normalizes the patch by cleaning it up (removing comments and other useless info).

**Comment by rb2k** [comment]
Should this be backported to 21.05?
I was just trying to package a nix release using:

```
nix-build release.nix -A binaryTarball.x86_64-darwin --arg nixpkgs 'fetchTarball channel:nixos-21.05'
```

and ran into this issue

**Comment by rb2k** [comment]
Actually, even with the patch applied, these URLs don't work anymore :)

**Pull request merged**